twitter.py

```diff
@@ -0,0 +1,165 @@
+                    "media_type" : f"image/{image_type}", # TODO support other types if necessary
+                    "total_bytes" : len(image_bytes),
+                    "media_category" : "tweet_image"
+                    }
+            files = {
+                    "media":image_bytes
+                    }
+            media_id = None
+            oauth = OAuth1Session(
+                self.consumer_key,
+                client_secret=self.consumer_secret,
+                resource_owner_key=self.access_token,
+                resource_owner_secret=self.access_secret,
+            )
+
+            response = oauth.post('https://upload.twitter.com/1.1/media/upload.json', json=request_data, files=files)
+            media_id = None
+            if not response.ok:
+                logtext = f"twitter media upload failed due to response {response.status_code} : {response.text}"
+                return None
+            else:
+                return str(response.json()["media_id"])
+        except Exception as e:
+            logtext = f("twitter media upload failed due to exception {e}")
+            return None
+        finally:
+            if logtext is not None:
+                if log is not None:
+                    log.error(logtext)
+                else:
+                    print(logtext)
+
+
+    def tweet(self, text, media_id=None, log=None):
+        # Input:
+        # text: A string containing the body of the tweet.
+        # media_id: A string containing a media_id for an image to be attached (e.g. the return value from an upload_media call).  If None then no image will be attached.
+        # Output:
+        # a boolean True/False indicating if the tweet was made successfully.
+
+        logtext = None
+        try:
+            # Make the request
+            oauth = OAuth1Session(
+                self.consumer_key,
+                client_secret=self.consumer_secret,
+                resource_owner_key=self.access_token,
+                resource_owner_secret=self.access_secret,
+            )
+
+            # Making the request
+            payload = {"text" : text}
+
+            # attach the media
+            if media_id:
+                payload["media"] = {"media_ids" : [media_id]}
+
+            response = oauth.post(
+                "https://api.twitter.com/2/tweets",
+                json=payload,
+            )
+
+            if response.status_code != 201:
+                logtext = f"tweet failed due to response {response.status_code} : {response.text}"
+                return False
+            return True
+        except Exception as e:
+            logtext = f"tweet failed due to exception {e}"
+            return False
+        finally:
+            if logtext is not None:
+                if log is not None:
+                    log.error(logtext)
+                else:
+                    print(logtext)
+
+if __name__ == "__main__":
+    twitter_login()
```

Comment:
any chance you can make this into an Admin page? E.g. the way there currently is a `/admin/authzoom` page that starts the zoom oauth process?

Reply:
Ah interesting, I’ll look into it!

Reply:
Hmm I think the twitter process is more manual that the zoom one.  I might be able to arrange something with an html form though.